presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractRowBlock.java

```diff
@@ -59,15 +60,16 @@ public RowBlockEncoding getEncoding()
     }
 
     @Override
-    public Block copyPositions(List<Integer> positions)
+    public Block copyPositions(int[] positions, int offset, int length)
     {
-        int newPositionCount = positions.size();
-        int[] newOffsets = new int[newPositionCount + 1];
-        boolean[] newRowIsNull = new boolean[newPositionCount];
+        checkValidPositionsArray(positions, offset, length);
 
-        List<Integer> fieldBlockPositions = new ArrayList<>(newPositionCount);
-        for (int i = 0; i < newPositionCount; i++) {
-            int position = positions.get(i);
+        int[] newOffsets = new int[length + 1];
+        boolean[] newRowIsNull = new boolean[length];
+
+        List<Integer> fieldBlockPositions = new ArrayList<>(length);
+        for (int i = 0; i < length; i++) {
```

Comment:
in other places we use zero based `0` index `i`

Reply:
I think the form @losipiuk suggested is simpler. How about changing all of them to that form?

Reply:
Initially all loops were indexed from `0` and I have changed selected loops to be indexed from `offset` by suggestion of @martint https://github.com/prestodb/presto/pull/8734#discussion_r151230849 and I agree with him on this one. I know it might be minor, but such loops are more efficient. I don't think java is smart enough to figure out that it can rewrite `for (int i = 0; i < length; i++) {` to `for (int i = offset; i < offset + length; ++i)` because it involves arithmetic that could potentially overflow. Therefore with `for (int i = 0; i < length; i++) {` you will end up with one unnecessary addition per row. I would rather leave it as it (I think both loops are simple enough) is instead of changing it back again.